File: java/medium/5_longest_palindromic_substring.java
// https://leetcode.com/problems/longest-palindromic-substring/

package java.medium;

class LongestPalindromicSubstring {
    private int maxLen = 0;

    public String longestPalindrome(String s) {
        int n = s.length();

        if (n < 2) {
            return s;
        }

        String result = "";

        for (int i = 0; i < n; i++) {
            // odd length -> single character center
            result = count(s, i, i, result);

            // even length -> consecutive character center
            result = count(s, i, i+1, result);

        }

        return result;
    }

    private String count(String s, int left, int right, String result) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            if (right - left + 1 > maxLen) {
                result = s.substring(left, right+1);
                maxLen = right - left + 1;
            }
            left--;
            right++;
        }

        return result;
    }
}

File: go/medium/5_longest_palindromic_substring.go
// https://leetcode.com/problems/longest-palindromic-substring/
package medium

func LongestPalindrome(s string) string {
	n := len(s)
	result := ""
	maxLen := 0

	var getLongest func(int, int)
	getLongest = func(l, r int) {
		for l >= 0 && r < n && s[l] == s[r] {
			if r-l+1 > maxLen {
				maxLen = r - l + 1
				result = s[l : r+1]
			}
			l--
			r++
		}
	}

	for i := 0; i < n; i++ {
		// odd length
		getLongest(i, i)

		// even length
		getLongest(i, i+1)
	}

	return result
}
